controller/contour/daemonset.go

```diff
@@ -61,25 +61,33 @@ const (
 	// envoyDaemonSetLabel identifies a daemonset as a contour daemonset,
 	// and the value is the name of the owning contour.
 	envoyDaemonSetLabel = "contour.operator.projectcontour.io/daemonset-envoy"
+	// xdsResourceVersion is the version of the Envoy xdS resource types.
+	xdsResourceVersion = "v3"
 )
 
 // ensureDaemonSet ensures a DaemonSet exists for the given contour.
-func (r *Reconciler) ensureDaemonSet(ctx context.Context, contour *operatorv1alpha1.Contour) error {
+func (r *Reconciler) ensureDaemonSet(ctx context.Context, contour *operatorv1alpha1.Contour) (*appsv1.DaemonSet, error) {
 	desired := DesiredDaemonSet(contour, r.Config.ContourImage, r.Config.EnvoyImage)
 
 	current, err := r.currentDaemonSet(ctx, contour)
 	if err != nil {
 		if errors.IsNotFound(err) {
-			return r.createDaemonSet(ctx, desired)
+			updated, err := r.createDaemonSet(ctx, desired)
+			if err != nil {
+				return nil, fmt.Errorf("failed to create daemonset for contour %s/%s: %w", contour.Namespace,
+					contour.Name, err)
+			}
+			return updated, nil
 		}
-		return fmt.Errorf("failed to get daemonset for contour %s/%s: %w", contour.Namespace, contour.Name, err)
+		return nil, fmt.Errorf("failed to get daemonset for contour %s/%s: %w", contour.Namespace, contour.Name, err)
 	}
 
-	if err := r.updateDaemonSetIfNeeded(ctx, current, desired); err != nil {
-		return fmt.Errorf("failed to update daemonset for contour %s/%s: %w", contour.Namespace, contour.Name, err)
+	updated, err := r.updateDaemonSetIfNeeded(ctx, current, desired)
```

Comment:
If the daemonset is actually updated here, should `ensureDaemonSet` call `currentDaemonSet` to return the resource living in the cluster _after_ the update?

Reply:
`updateDaemonSetIfNeeded()` updates the ds on the api server if `current` doesn't match `desired`. If `updateDaemonSetIfNeeded()` returns an error value, reconciliation of the `Contour` will be triggered and the reconciler will get the latest `Contour` from the api server. Otherwise `updateDaemonSetIfNeeded()` returns a nil error value and `ensureDaemonSet` completes successfully.